core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala

```diff
@@ -208,6 +201,87 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
       query.as[CompactEntityRefRecord]
     }
 
+  def copyEntitiesToNewWorkspace(sourceWs: UUID,
+                                 destWs: UUID,
+                                 entityRefs: Set[AttributeEntityReference] = Set()
+                                ): WriteAction[Int] = {
+
+    def copyChunkOfEntitiesOrAllEntities(chunk: Set[AttributeEntityReference] = Set()) =
+      for {
+        entitiesCopiedCount <- copyEntities(sourceWs, destWs, chunk)
+      } yield entitiesCopiedCount
+
+      val chunks: Iterator[Set[AttributeEntityReference]] = if (entityRefs.size > driverComponent.batchSize) {
+        entityRefs.grouped(driverComponent.batchSize)
+      } else {
+        Iterator(entityRefs)
+      }
+
+    val allCopies = DBIO.sequence(chunks map copyChunkOfEntitiesOrAllEntities)
+
+    allCopies.map { copyActionResults: Iterator[Int] =>
+      copyActionResults.sum
+    }
+  }
+
+  def getEntitySubtrees(workspaceId: UUID,
+                        entityType: String,
+                        entityNames: Set[String]
+                       ): ReadAction[Seq[EntityPath]] = {
+    val refs = entityNames.map { name => AttributeEntityReference(entityType, name) }
+    for {
+      startingEntityRecords <- getEntityRefs(workspaceId, refs)
+      entities = startingEntityRecords.map { record => record.toAttributeEntityReference}
+      refsToId = startingEntityRecords.map(record => EntityPath(Seq(record.toAttributeEntityReference)) -> record.toAttributeEntityReference).toMap
+      allRefs <- recursiveGetEntityReferences(workspaceId, entities.toSet, refsToId)
+    } yield allRefs
+  }
+
+  def recursiveGetEntityReferences(workspaceId: UUID,
```

Comment:
I think you should try a recursive query https://dev.mysql.com/doc/refman/9.0/en/with.html#common-table-expressions-recursive

Reply:
you could try using AI with a prompt like "reimplement recursiveGetEntityReferences using a recursive CTE"

Reply:
+1 to using a recursive CTE; I think it should be feasible against the `ENTITY_REFS` table